utils/self-extr-exec/decompressor.cpp

```diff
@@ -0,0 +1,374 @@
+void fillCommand(char command[], int argc, char * argv[], size_t length)
+{
+    memset(command, '\0', 3 + strlen(argv[0]) + 14 + strlen(argv[0]) + 4 + strlen(argv[0]) + length + argc);
+
+    /// position in command
+    size_t shift = 0;
+
+    /// Support variables to create command
+    char mv[] = "mv ";
+    char decompressed[] = ".decompressed ";
+    char add_command[] = " && ";
+    char space[] = " ";
+
+    fill(command, mv, 3, shift);
+    fill(command, argv[0], strlen(argv[0]), shift);
+    fill(command, decompressed, 14, shift);
+    fill(command, argv[0], strlen(argv[0]), shift);
+    fill(command, add_command, 4, shift);
+    fill(command, argv[0], strlen(argv[0]), shift);
+    fill(command, space, 1, shift);
+
+    /// forward all arguments
+    for (int i = 1; i < argc; ++i)
+    {
+        fill(command, argv[i], strlen(argv[i]), shift);
+        if (i != argc - 1)
+            fill(command, space, 1, shift);
+    }
+}
+
+int main(int argc, char* argv[])
+{
+    char file_path[strlen(argv[0]) + 1];
+    memset(file_path, 0, sizeof(file_path));
+    strcpy(file_path, argv[0]);
+
+    char * path = nullptr;
+    char * name = strrchr(file_path, '/');
+    if (name)
+    {
+        path = file_path;
+        *name = 0;
+        ++name;
+    }
+    else
+        name = file_path;
+
+    int input_fd = open(argv[0], O_RDONLY);
+    if (input_fd == -1)
+    {
+        perror(nullptr);
+        return 1;
+    }
+
+    bool have_compressed_analoge = false;
+
+    /// Decompress all files
+    if (0 != decompressFiles(input_fd, path, name, have_compressed_analoge))
+    {
+        printf("Error happened during decompression.\n");
+        if (0 != close(input_fd))
+            perror(nullptr);
+        return 1;
+    }
+
+    if (0 != close(input_fd))
+        perror(nullptr);
+
+    /// According to documentation `mv` will rename file if it
+    /// doesn't move to other directory.
+    /// Sometimes `rename` doesn't exist by default and
+    /// `rename.ul` is set instead. It will lead to errors
+    /// that can be easily avoided with help of `mv`
+
+    if (!have_compressed_analoge)
+    {
+        printf("No target executable - decompression only was performed.\n");
+        /// remove file
+        execlp("rm", "rm", argv[0], NULL);
+        perror(nullptr);
+        return 1;
+    }
+    else
+    {
+        /// move decompressed file instead of this binary and apply command
+        char bash[] = "sh";
+        char executable[] = "-c";
+
+        /// length of forwarded args
+        size_t length = 0;
+        for (int i = 1; i < argc; ++i)
+            length += strlen(argv[i]);
+
+        /// mv filename.decompressed filename && filename agrs...
+        char command[3 + strlen(argv[0]) + 14 + strlen(argv[0]) + 4 + strlen(argv[0]) + length + argc];
+        fillCommand(command, argc, argv, length);
+
+        /// replace file and call executable
+        char * newargv[] = { bash, executable, command, nullptr };
```

Comment:
Why do we run it with `sh -c`?
Why is shell needed?

It makes everything more complicated, and proper argument forwarding is nearly impossible.

I think ClickHouse should run under chroot with no shell.

Reply:
as I remember the reason for using sh was that attempt to run just clickhouse gives "text file is busy" error.
And what about chroot here? how it helps?

Reply:
You can workaround the "text file busy" error by renaming the old executable to a temporary file first.
Chroot will not help, it is just a motivation to not use `sh`.

Reply:
will try

Reply:
@alexey-milovidov doesn't work either. Algorithm is next:
```
system("mv " + argv[0] + " " + argv[0] + ".tmp");
system("mv " + argv[0] + ".decompressed " + argv[0]);
system("rm " + argv[0] + ".tmp"); // optional, result is the same without it
execvp(argv[0], argv);
```
result the same - last execvp gives "Text file busy" error

Reply:
@alexey-milovidov BUT replacing `mv clickhouse.decompressed clickhouse` by `cp clickhouse.decompressed clickhouse` seems to work! :)

Reply:
@alexey-milovidov https://github.com/ClickHouse/ClickHouse/pull/39763